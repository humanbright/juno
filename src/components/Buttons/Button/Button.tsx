import MuiButton from '@material-ui/core/Button';
import React, {
  ComponentProps,
  forwardRef,
  useCallback,
  useLayoutEffect,
  useMemo,
  useRef,
} from 'react';

import {
  combineClasses,
  isRcElement,
  RcBaseColor,
  RcBaseProps,
  RcPaletteKeys,
  RcTheme,
  removeClassName,
  styled,
  useForkRef,
  useTheme,
  useThemeProps,
  withDeprecatedCheck,
} from '../../../foundation';
import { RcIcon, RcIconProps } from '../../Icon';
import { RcIconSizes } from '../../Icon/utils';
import { RcCircularProgress, RcCircularProgressProps } from '../../Progress';
import { withTooltip } from '../../Tooltip';
import { buttonStyle } from './styles';
import { getButtonIconSize, RcButtonClasses, RcButtonSize } from './utils';

type RcButtonColor = RcBaseColor<
  'primary' | 'secondary' | 'negative' | 'positive' | 'action' | 'neutral'
>;

type RcButtonColorLoadingMode = 'prefix' | 'replace' | 'suffix';

type RcButtonVariant = 'text' | 'outlined' | 'contained' | 'plain';

type RcButtonRadius = keyof RcTheme['radius'];

type RcButtonProps = {
  /** size of button, default is `large` */
  size?: RcButtonSize;
  /** variant of button, default is `contained` */
  variant?: RcButtonVariant;
  /** color of button */
  color?: RcPaletteKeys | RcButtonColor;
  /** is button loading, when `loading` that default `disabled` will be true  */
  loading?: boolean;
  /** loading mode with button, default is `replace` */
  loadingMode?: RcButtonColorLoadingMode;
  /** Props send to `RcCircularProgress` when loading is `true` */
  CircularProgressProps?: RcCircularProgressProps;
  /** component for render root button, default is `button` */
  component?: React.ElementType;
  /** border radius for button */
  radius?: RcButtonRadius;
  /** should keep elevation when type is `container` */
  keepElevation?: boolean;
  /** @deprecated Icon, please use startIcon with `<RcIcon />` */
  IconProps?: RcIconProps;
} & RcBaseProps<
  ComponentProps<typeof MuiButton>,
  | 'color'
  | 'size'
  | 'variant'
  | 'title'
  // * use disableFocusRipple, so omit that
  | 'focusRipple'
>;

const _RcButton = forwardRef<any, RcButtonProps>(
  (inProps: RcButtonProps, ref) => {
    const props = useThemeProps({ props: inProps, name: 'RcButton' });
    const {
      children: childrenProp,
      classes: classesProp,
      startIcon: startIconProp,
      endIcon: endIconProp,
      loading,
      loadingMode,
      disabled = loading,
      IconProps,
      size,
      color,
      CircularProgressProps,
      variant,
      radius,
      keepElevation,
      ...rest
    } = props;

    const theme = useTheme();
    const innerRef = useRef<HTMLButtonElement>(null);
    const buttonRef = useForkRef(innerRef, ref);

    const isPlain = variant === 'plain';
    const isReplace = loadingMode === 'replace';
    const iconSize = getButtonIconSize(size);

    const loadingElm = useMemo(() => {
      if (loading) {
        const progressSize = RcIconSizes[iconSize];

        return (
          <RcCircularProgress
            size={progressSize}
            color="inherit"
            {...CircularProgressProps}
          />
        );
      }
      return undefined;
    }, [CircularProgressProps, iconSize, loading]);

    const getCurrIcon = useCallback(
      (currIcon: React.ReactNode) => {
        if (
          isRcElement<RcIconProps>(currIcon, ['RcIcon']) &&
          currIcon.props.size === RcIcon.defaultProps!.size
        ) {
          return React.cloneElement(currIcon, { size: iconSize });
        }

        return currIcon;
      },
      [iconSize],
    );

    const startIcon = useMemo(() => {
      if (loadingElm) {
        switch (loadingMode) {
          case 'prefix':
            return loadingElm;
          case 'replace':
            return undefined;
          default:
            break;
        }
      }

      if (startIconProp) {
        return getCurrIcon(startIconProp);
      }

      // TODO: IconProps will be remove
      if (IconProps) {
        return <RcIcon size={iconSize} {...IconProps} />;
      }

      return undefined;
    }, [
      IconProps,
      getCurrIcon,
      iconSize,
      loadingElm,
      loadingMode,
      startIconProp,
    ]);

    const endIcon = useMemo(() => {
      if (loadingElm) {
        switch (loadingMode) {
          case 'suffix':
            return loadingElm;
          case 'replace':
            return undefined;
          default:
            break;
        }
      }

      if (endIconProp) {
        return getCurrIcon(endIconProp);
      }

      return undefined;
    }, [loadingElm, endIconProp, loadingMode, getCurrIcon]);

    const classes = useMemo(
      () => combineClasses(RcButtonClasses, classesProp),
      [classesProp],
    );

    // * need remove not need MuiButton-iconSizeMedium for icon render correct
    useLayoutEffect(() => {
      removeClassName(innerRef, 'MuiButton-iconSizeMedium');
    });

    return (
      <MuiButton
        ref={buttonRef}
        disabled={disabled}
        variant={isPlain ? undefined : (variant as any)}
        disableRipple={theme?.props?.MuiButton?.disableRipple || isPlain}
        startIcon={startIcon}
        endIcon={endIcon}
        classes={classes}
        {...rest}
      >
        {loading && isReplace ? loadingElm : childrenProp}
      </MuiButton>
    );
  },
);

/** @release */
const RcButton = styled(
  withDeprecatedCheck(
    withTooltip(_RcButton),
    [
      {
        prop: 'IconProps',
        time: '2021-4',
        comment: `@deprecated Icon, please use startIcon with \`<RcIcon />\` */`,
      },
    ],
    'RcButton',
  ),
)`
  ${buttonStyle}
`;

RcButton.defaultProps = {
  size: 'large',
  color: 'primary',
  variant: 'contained',
  loadingMode: 'replace',
};

RcButton.displayName = 'RcButton';

export {
  RcButton,
  RcButtonSize,
  RcButtonProps,
  RcButtonColor,
  RcButtonColorLoadingMode,
  RcButtonVariant,
};
