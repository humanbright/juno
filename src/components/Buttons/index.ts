export * from './Button';
export * from './ButtonBar';
export * from './ButtonBase';
export * from './ButtonGroup';
export * from './FabButton';
export * from './FabIconButton';
export * from './IconButton';
export * from './IconButtonGroup';
export * from './RoundButton';
export * from './SplitButton';
export * from './ToggleButton';
export * from './ToggleButtonGroup';
