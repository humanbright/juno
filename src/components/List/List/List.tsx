import MuiList from '@material-ui/core/List';
import React, { ComponentProps, forwardRef, useMemo } from 'react';

import {
  combineClasses,
  RcBaseProps,
  styled,
  useThemeProps,
} from '../../../foundation';
import { ListStyle } from './styles';
import { RcListClasses } from './utils';

type RcListProps = {} & RcBaseProps<ComponentProps<typeof MuiList>>;

const _RcList = forwardRef<any, RcListProps>((inProps, ref) => {
  const props = useThemeProps({ props: inProps, name: 'RcList' });
  const { classes: classesProp, children, ...rest } = props;
  const classes = useMemo(
    () => combineClasses(RcListClasses, classesProp),
    [classesProp],
  );

  return (
    <MuiList {...rest} ref={ref} classes={classes}>
      {children}
    </MuiList>
  );
});

const RcList = styled(_RcList)`
  ${ListStyle}
`;

RcList.defaultProps = {
  disablePadding: true,
};

RcList.displayName = 'RcList';

export { RcList };
export type { RcListProps };
