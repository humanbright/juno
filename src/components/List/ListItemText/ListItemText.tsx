import MuiListItemText from '@material-ui/core/ListItemText';
import React, { ComponentProps, forwardRef, useMemo } from 'react';

import {
  combineClasses,
  omit,
  RcBaseProps,
  RcPaletteProp,
  styled,
  withDeprecatedCheck,
  useThemeProps,
} from '../../../foundation';
import { RcTypographyProps } from '../../Typography';
import { ListItemTextStyle } from './styles';
import { RcListItemTextClasses } from './utils';

type TypographyProps = RcBaseProps<RcTypographyProps, 'weight'>;

type RcListItemTextProps = {
  /** is text Ellipsis, default is true. */
  isEllipsis?: boolean;
  /** These props will be forwarded to the primary typography component (as long as disableTypography is not true). */
  primaryTypographyProps?: TypographyProps;
  /** These props will be forwarded to the secondary typography component (as long as disableTypography is not true). */
  secondaryTypographyProps?: TypographyProps;
  /** @deprecated define color for primary, recommend use primaryTypographyProps color prop */
  primaryColor?: RcPaletteProp;
  /** @deprecated is alignCenter, recommend using classes to define it. */
  alignCenter?: boolean;
  /** @deprecated is lineThrough, recommend using classes to define it. */
  lineThrough?: boolean;
} & RcBaseProps<
  ComponentProps<typeof MuiListItemText>,
  'primaryTypographyProps' | 'secondaryTypographyProps'
>;

const _RcListItemText = forwardRef<any, RcListItemTextProps>((inProps, ref) => {
  const props = useThemeProps({ props: inProps, name: 'RcListItemText' });
  const {
    classes: classesProp,
    children,
    isEllipsis,
    primaryTypographyProps: primaryTypographyPropsProp,
    secondaryTypographyProps: secondaryTypographyPropsProp,
    primaryColor,
    alignCenter,
    lineThrough,
    ...rest
  } = props;

  const classes = useMemo(
    () => combineClasses(RcListItemTextClasses, classesProp),
    [classesProp],
  );

  const primaryTypographyProps = useMemo(
    () => omit(primaryTypographyPropsProp, ['color', 'variant']),
    [primaryTypographyPropsProp],
  );

  const secondaryTypographyProps = useMemo(
    () => omit(secondaryTypographyPropsProp, ['color', 'variant']),
    [secondaryTypographyPropsProp],
  );

  return (
    <MuiListItemText
      {...rest}
      ref={ref}
      classes={classes}
      primaryTypographyProps={primaryTypographyProps}
      secondaryTypographyProps={secondaryTypographyProps}
    >
      {children}
    </MuiListItemText>
  );
});

const RcListItemText = styled(
  withDeprecatedCheck(
    _RcListItemText,
    [
      {
        prop: 'primaryColor',
        time: '2021-4',
        comment: 'recommend use primaryTypographyProps color prop',
      },
      {
        prop: 'alignCenter',
        time: '2021-4',
        comment: `recommend using classes to define it`,
      },
      {
        prop: 'lineThrough',
        time: '2021-4',
        comment: `recommend using classes to define it`,
      },
    ],
    'RcListItemText',
  ),
)`
  ${ListItemTextStyle}
`;

RcListItemText.defaultProps = {
  isEllipsis: true,
  primaryTypographyProps: {
    variant: 'body1',
    color: 'neutral.f06',
  },
  secondaryTypographyProps: {
    variant: 'caption1',
    color: 'neutral.f04',
  },
};

RcListItemText.displayName = 'RcListItemText';

export { RcListItemText };
export type { RcListItemTextProps };
