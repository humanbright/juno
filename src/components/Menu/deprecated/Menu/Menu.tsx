import MuiMenu from '@material-ui/core/Menu';
import React, {
  ComponentProps,
  forwardRef,
  memo,
  useMemo,
  useState,
} from 'react';

import {
  RcBaseProps,
  styled,
  useEventCallback,
  useRcPortalWindowContext,
  withDeprecatedCheck,
} from '../../../../foundation';
import { StyledMenu } from './styles';

type RcMenuProps = {
  autoClose?: boolean;
  /** @deprecated use autoClose to replace that */
  enableCloseMenuWhenItemClicked?: boolean;
} & RcBaseProps<ComponentProps<typeof MuiMenu>>;

type CloseReasonType = 'backdropClick' | 'escapeKeyDown' | 'tabKeyDown';

type MenuContextType = {
  autoClose: boolean;
  open: boolean;
  closeMenu: (event: {}, reason: CloseReasonType) => void;
  focusedMenuItemId: string;
  setFocusedMenuItemId: (id: string) => void;
};

const MenuContext = React.createContext<MenuContextType>({
  autoClose: false,
  open: false,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  closeMenu: (event: {}, reason: CloseReasonType) => {},
  focusedMenuItemId: '',
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  setFocusedMenuItemId: (id: string) => {},
});

// TODO: should use ref to refactor this menu logic with focus item set
const MemoMenu = memo(StyledMenu);

const _RcMenu = forwardRef<any, RcMenuProps>(
  (
    {
      onClose,
      open,
      enableCloseMenuWhenItemClicked,
      autoClose = enableCloseMenuWhenItemClicked,
      children,
      ...rest
    },
    ref,
  ) => {
    const [focusedMenuItemId, setFocusedMenuItemId] = useState('');
    const { externalWindow } = useRcPortalWindowContext();

    const handleClose = useEventCallback(
      (event: {}, reason: CloseReasonType) => {
        onClose?.(event, reason as Exclude<CloseReasonType, 'tabKeyDown'>);
      },
    );

    const ctxValue: MenuContextType = useMemo(
      () => ({
        autoClose: !!autoClose,
        open,
        closeMenu: handleClose,
        focusedMenuItemId,
        setFocusedMenuItemId: (id: string) => setFocusedMenuItemId(id),
      }),
      [autoClose, focusedMenuItemId, handleClose, open],
    );

    return (
      <MenuContext.Provider value={ctxValue}>
        <MemoMenu
          container={externalWindow?.document.body}
          {...rest}
          open={open}
          ref={ref}
          onClose={handleClose}
          disableEnforceFocus
        >
          {children}
        </MemoMenu>
      </MenuContext.Provider>
    );
  },
);

/** @release */
const RcMenu = styled(
  withDeprecatedCheck(
    _RcMenu,
    [
      {
        prop: 'enableCloseMenuWhenItemClicked',
        time: '2021-2',
        comment: `use autoClose to replace that `,
      },
    ],
    'RcMenu',
  ),
)``;

RcMenu.displayName = 'RcMenu';

RcMenu.defaultProps = {
  getContentAnchorEl: null,
  anchorOrigin: { vertical: 'bottom', horizontal: 'left' },
};

export { RcMenu, RcMenuProps, MenuContext as RcMenuContext };
