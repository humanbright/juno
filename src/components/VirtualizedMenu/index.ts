export * from './VirtualizedMenu';
export * from './VirtualizedMenuList';
export * from './VirtualizedDivider';
