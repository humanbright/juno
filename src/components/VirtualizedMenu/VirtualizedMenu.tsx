import clsx from 'clsx';
import React, { forwardRef, useMemo, useRef } from 'react';

import {
  combineClasses,
  combineProps,
  RcBaseProps,
  styled,
  useEventCallback,
  useForkRef,
  useRcPortalWindowContext,
  useThemeProps,
} from '../../foundation';
import { RcMenuProps } from '../Menu';
import { RcPopover } from '../Popover';
import { VirtualizedMenuStyle } from './styles';
import { RcVirtualizedMenuClasses } from './utils';
import {
  RcVirtualizedMenuList,
  RcVirtualizedMenuListProps,
  RcVirtualizedMenuListRef,
} from './VirtualizedMenuList';

type RcVirtualizedMenuProps = {
  /** props for apply on inner `MenuList` */
  MenuListProps?: RcVirtualizedMenuListProps;
} & RcBaseProps<RcMenuProps, 'MenuListProps'>;

const _RcVirtualizedMenu = forwardRef<any, RcVirtualizedMenuProps>(
  (inProps, ref) => {
    const props = useThemeProps({ props: inProps, name: 'RcVirtualizedMenu' });
    const {
      autoFocus = true,
      children,
      classes: classesProp,
      disableAutoFocusItem = false,
      MenuListProps: { maxHeight = 416, onKeyDown, ...MenuListProps } = {},
      onClose,
      open,
      onEntering,
      onEntered,
      PaperProps: PaperPropsProp = {},
      PopoverClasses,
      transitionDuration = 'auto',
      variant = 'selectedMenu',
      ...rest
    } = props;
    const { document } = useRcPortalWindowContext();

    const popoverRef = useRef<HTMLDivElement>(null);
    const handleRef = useForkRef(ref, popoverRef);

    const menuListActionRef = useRef<RcVirtualizedMenuListRef>(null);

    const classes = useMemo(
      () => combineClasses(RcVirtualizedMenuClasses, classesProp),
      [classesProp],
    );

    const autoFocusItem = autoFocus && !disableAutoFocusItem && open;

    const events = useMemo(
      () =>
        combineProps(
          {
            onEntering: (element: HTMLElement, isAppearing: boolean) => {
              menuListActionRef.current?.adjustStyleForScrollbar();
              menuListActionRef.current?.scrollIntoViewAndFocus();

              onEntering?.(element, isAppearing);
            },
            onEntered: () => {
              const popoverElm = popoverRef.current!;
              const paperElm = popoverElm.querySelector(
                `.${RcVirtualizedMenuClasses.paper}`,
              )!;

              // check that paper height is same as vl scroll, if not adjust that containerHeight
              const newHeight = paperElm.clientHeight;
              if (newHeight !== maxHeight) {
                menuListActionRef.current?.adjustContainerHeight(newHeight);
              }
            },
          },
          {
            onEntering,
            onEntered,
          },
        ),
      [maxHeight, onEntered, onEntering],
    );

    const PaperProps = combineProps(
      {
        classes: {
          root: classes?.paper,
        },
      },
      PaperPropsProp,
    );

    const handleListKeyDown = useEventCallback(
      (event: React.KeyboardEvent<HTMLUListElement>): void => {
        onKeyDown?.(event);

        if (event.key === 'Tab') {
          event.preventDefault();

          if (onClose) {
            onClose(event, 'tabKeyDown');
          }
        }
      },
    );

    return (
      <RcPopover
        ref={handleRef}
        container={document.body}
        classes={PopoverClasses}
        onClose={onClose}
        open={open}
        transitionDuration={transitionDuration}
        PaperProps={PaperProps}
        {...events}
        {...rest}
      >
        <RcVirtualizedMenuList
          position="unset"
          action={menuListActionRef}
          autoFocus={autoFocus}
          autoFocusItem={autoFocusItem}
          maxHeight={maxHeight}
          variant={variant}
          onKeyDown={handleListKeyDown}
          {...MenuListProps}
          className={clsx(classes.list, MenuListProps.className)}
        >
          {children}
        </RcVirtualizedMenuList>
      </RcPopover>
    );
  },
);

const RcVirtualizedMenu = styled(_RcVirtualizedMenu)`
  ${VirtualizedMenuStyle}
`;

RcVirtualizedMenu.defaultProps = {
  autoFocus: true,
  variant: 'selectedMenu',
  getContentAnchorEl: null,
  anchorReference: 'anchorEl',
  anchorOrigin: { vertical: 'bottom', horizontal: 'left' },
};

export { RcVirtualizedMenu, RcVirtualizedMenuProps };
