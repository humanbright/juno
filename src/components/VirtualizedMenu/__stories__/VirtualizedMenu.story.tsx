import { Meta, Story } from '@storybook/react/types-6-0';
import range from 'lodash/range';
import React, { ComponentProps, FunctionComponent, useState } from 'react';

import {
  notControlInDocTable,
  notShowInDocTable,
  sortInDocTable,
  switchToControlKnobs,
} from '../../../storybook';
import { RcButton } from '../../Buttons/Button';
import { RcListItemText } from '../../List/ListItemText';
import { RcMenuItem } from '../../Menu/MenuItem';
import { RcVirtualizedMenu } from '../VirtualizedMenu';

export default {
  title: 'ðŸš€ Cleanup Components/VirtualizedMenu',
  component: RcVirtualizedMenu,
  excludeStories: /MenuExampleComponent/,
  argTypes: {
    ...sortInDocTable<keyof MenuProps>([
      // Menu
      'open',
      'anchorEl',
      'autoFocus',
      'autoClose',
      'variant',
      'onClose',
      'classes',
      'keepMounted', // Modal

      // MenuList
      'MenuListProps',
      'disableAutoFocusItem',
      'disableAutoFocus', // Modal

      // Popover
      'action',
      'anchorOrigin',
      'anchorPosition',
      'anchorReference',
      'elevation',
      'getContentAnchorEl',
      'marginThreshold',
      'PaperProps',
      'PopoverClasses',
      'TransitionProps',
      'TransitionComponent',
      'transformOrigin',
      'transitionDuration',

      // Modal
      'BackdropComponent',
      'BackdropProps',
      'closeAfterTransition',
      'container',
      'disableBackdropClick',
      // 'disableEnforceFocus',
      'disableEscapeKeyDown',
      'disablePortal',
      'disableRestoreFocus',
      'disableScrollLock',
      'hideBackdrop',
      'onBackdropClick',
      'onEscapeKeyDown',
      'onRendered',
    ]),
    ...notControlInDocTable<keyof MenuProps>([
      'open',
      'anchorEl',
      'anchorOrigin',
    ]),
    ...notShowInDocTable<keyof MenuProps>([]),
  },
} as Meta;

type MenuProps = ComponentProps<typeof RcVirtualizedMenu>;

const menus = range(0, 1000);

export const VirtualizedMenu: Story<Partial<MenuProps>> = ({
  children,
  anchorOrigin,
  ...args
}) => {
  switchToControlKnobs();

  const [anchorEl, setAnchorEl] = useState<HTMLButtonElement | null>(null);
  const [count, setCount] = useState(0);
  const handleClick = (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>,
  ) => setAnchorEl(event.currentTarget);

  const handleClose = () => setAnchorEl(null);

  return (
    <>
      <RcButton onClick={handleClick}>Open Menu</RcButton>
      <RcVirtualizedMenu
        {...args}
        open={Boolean(anchorEl)}
        anchorEl={anchorEl}
        // MenuListProps={{ disabledItemsFocusable: true }}
        onClose={handleClose}
      >
        {menus.map((x) => {
          const label = `${x}-MenuItem${count}`;
          return (
            <RcMenuItem
              key={x}
              // disabled={x < 10}
              onClick={handleClose}
              onContextMenu={() => {
                setCount(count + 1);
              }}
              data-search-text={label}
            >
              <RcListItemText primary={label} />
            </RcMenuItem>
          );
        })}
      </RcVirtualizedMenu>
    </>
  );
};

VirtualizedMenu.args = {};

VirtualizedMenu.argTypes = {
  ...notControlInDocTable<keyof MenuProps>([]),
};

VirtualizedMenu.parameters = {
  tags: [
    {
      name: 'Mui',
      href: 'https://material-ui.com/components/menus/',
      value: 'Menu',
    },
    {
      name: 'Source',
      value: 'virtuoso',
      href: 'https://virtuoso.dev/',
      color: 'avatar.lake',
    },
    {
      name: 'Accessibility',
      value: '100%',
    },
  ],
};

type MenuExampleComponentProps = {} & Partial<
  ComponentProps<typeof RcVirtualizedMenu>
>;

export const MenuExampleComponent: FunctionComponent<MenuExampleComponentProps> = (
  props,
) => {
  const [anchorEl, setAnchorEl] = useState(null);

  const handleClick = (event: any) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  return (
    <>
      <RcButton onClick={handleClick}>Open Menu</RcButton>
      <RcVirtualizedMenu
        {...props}
        anchorEl={anchorEl}
        open={Boolean(anchorEl)}
        keepMounted
        onClose={handleClose}
      >
        {menus.map((x) => (
          <RcMenuItem key={x} onClick={handleClose}>
            <RcListItemText primary={`MenuItem ${x}`} />
          </RcMenuItem>
        ))}
        <RcMenuItem key="cool">MenuItem Five</RcMenuItem>
      </RcVirtualizedMenu>
    </>
  );
};

export const MenuExamples: Story<MenuProps> = ({ anchorOrigin, ...args }) => {
  switchToControlKnobs();

  return <MenuExampleComponent {...args} />;
};

MenuExamples.argTypes = {
  ...notControlInDocTable<keyof MenuProps>(['anchorOrigin']),
};
