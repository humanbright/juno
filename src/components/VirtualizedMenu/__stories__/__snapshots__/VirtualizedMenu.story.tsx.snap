// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots ðŸš€ Cleanup Components/VirtualizedMenu Menu Examples 1`] = `
Array [
  .c0 {
  text-transform: none;
  font-size: 0.9375rem;
  font-weight: 700;
  font-family: Lato,Helvetica,Arial,sans-serif;
  text-align: center;
  box-shadow: unset;
  min-width: 96px;
  min-height: 36px;
  padding: 0px 16px;
}

.c0 .RcButton-startIcon {
  margin-right: 8px;
}

.c0 .RcButton-endIcon {
  margin-left: 8px;
}

.c0.RcButton-text {
  color: #066FAC;
}

.c0.RcButton-contained {
  color: #fff;
  background-color: #066FAC;
}

.c0.RcButton-contained.RcButton-disabled {
  background-color: #B9B9B9;
  color: #FFFFFF;
}

.c0.RcButton-outlined {
  color: #066FAC;
  border-color: #066FAC;
  padding: 0px 15px;
}

@media (hover:hover) and (pointer:fine) {
  .c0.RcButton-text:hover {
    background-color: rgba(6,111,172,0.08);
  }
}

@media (hover:hover) and (pointer:fine) {
  .c0.RcButton-contained:hover {
    background-color: rgba(6,111,172,0.92);
  }
}

@media (hover:hover) and (pointer:fine) {
  .c0.RcButton-outlined:hover {
    background-color: rgba(6,111,172,0.08);
  }
}

<button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained RcButton-contained c0"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Open Menu
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>,
  .c1 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c2 {
  height: 8px;
}

.c0 .RcVirtualizedMenu-paper {
  max-height: calc(100% - 96px);
  -webkit-overflow-scrolling: touch;
  overflow: hidden;
}

@media (hover:hover) and (pointer:fine) {

}

@media (hover:hover) and (pointer:fine) {

}

@media (hover:hover) and (pointer:fine) {

}

<div
    className="MuiPopover-root c0"
    onKeyDown={[Function]}
    role="presentation"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "fixed",
        "right": 0,
        "top": 0,
        "visibility": "hidden",
        "zIndex": 1300,
      }
    }
  >
    <div
      data-test="sentinelStart"
      tabIndex={0}
    />
    <div
      className="MuiPaper-root RcVirtualizedMenu-paper MuiPopover-paper MuiPaper-elevation8 MuiPaper-rounded"
      style={
        Object {
          "opacity": 0,
          "transform": "scale(0.75, 0.5625)",
          "visibility": "hidden",
        }
      }
      tabIndex="-1"
    >
      <span
        className="c1"
        onFocus={[Function]}
      />
      <div
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "WebkitOverflowScrolling": "touch",
            "height": 416,
            "outline": "none",
            "overflowY": "auto",
            "position": "relative",
          }
        }
        tabIndex={-1}
      >
        <div
          style={
            Object {
              "height": "100%",
              "position": "absolute",
              "top": 0,
              "width": "100%",
            }
          }
        >
          <div>
            <div
              className="c2"
            />
          </div>
          <ul
            className="MuiList-root "
            onKeyDown={[Function]}
            role="menu"
            style={
              Object {
                "boxSizing": "border-box",
                "marginTop": 0,
                "paddingBottom": 0,
                "paddingTop": 0,
              }
            }
          />
          <div>
            <div
              className="c2"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabIndex={0}
    />
  </div>,
]
`;

exports[`Storyshots ðŸš€ Cleanup Components/VirtualizedMenu Virtualized Menu 1`] = `
.c0 {
  text-transform: none;
  font-size: 0.9375rem;
  font-weight: 700;
  font-family: Lato,Helvetica,Arial,sans-serif;
  text-align: center;
  box-shadow: unset;
  min-width: 96px;
  min-height: 36px;
  padding: 0px 16px;
}

.c0 .RcButton-startIcon {
  margin-right: 8px;
}

.c0 .RcButton-endIcon {
  margin-left: 8px;
}

.c0.RcButton-text {
  color: #066FAC;
}

.c0.RcButton-contained {
  color: #fff;
  background-color: #066FAC;
}

.c0.RcButton-contained.RcButton-disabled {
  background-color: #B9B9B9;
  color: #FFFFFF;
}

.c0.RcButton-outlined {
  color: #066FAC;
  border-color: #066FAC;
  padding: 0px 15px;
}

@media (hover:hover) and (pointer:fine) {
  .c0.RcButton-text:hover {
    background-color: rgba(6,111,172,0.08);
  }
}

@media (hover:hover) and (pointer:fine) {
  .c0.RcButton-contained:hover {
    background-color: rgba(6,111,172,0.92);
  }
}

@media (hover:hover) and (pointer:fine) {
  .c0.RcButton-outlined:hover {
    background-color: rgba(6,111,172,0.08);
  }
}

<button
  className="MuiButtonBase-root MuiButton-root MuiButton-contained RcButton-contained c0"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    Open Menu
  </span>
</button>
`;
